The Pattern-Matching Bug: remove `check_partial`

gasche opened the pull request:
(This PR sits on top of #13152, and is thus best review by looking at the last commit only.)

This is a non-obvious code simplification PR after #7241 is fixed. Before #7241 was reported and worked on, there already was a heuristic named `check_partial` to degrade pattern-matching compilation in some cases involving side-effects during matching. Now that #7241 is fixed by other changes in the pattern-matching compiler, the heuristic is not necessary anymore and complicates the code, it can be removed.

The testsuite `testsuite/tests/match-side-effects/check_partial` was introduced in #13121 to test various cases where the `check_partial` heuristic would fire. The code generated for these tests is *not* modified by the present PR, demonstrating that, on those tests, removing `check_partial` does not change the compiler behavior.

### Details

The `check_partial` heuristics was a coarse-grained approach to degrade matching totality information in certain cases -- when clauses contain both reads of mutable fields and either guards or forcing of lazy patterns.

This heuristic is unfortunately not quite correct (it is not enough to prevent https://github.com/ocaml/ocaml/issues/7241), and is not sufficient conceptually for OCaml Multicore. In a Multicore world, other domains may race to update values concurrently, so we cannot assume that only a fixed set of matching constructions may result in side-effects.

gasche committed:
matching: remove the [check_partial] heuristics

The [check_partial] heuristics are a coarse-grained approach degrade
matching totality information in certain cases -- when clauses contain
both reads of mutable fields and either guards or forcing of lazy
patterns.

It is not quite correct (it is not enough to prevent #7241), and is
not sufficient conceptually for OCaml Multicore. In a Multicore world,
other domains may race to update values concurrently, so we cannot
assume that only a fixed set of matching constructions may result in
side-effects.

This heuristic is subsumed by the recent changes to:
  - make context information accurate by binding mutable fields eagerly
  - make totality information accurate by tracking the mutability of
    the current position
and can now be retired.

Note: the [check_partial] function also contained hidden logic to deal
with the pattern-matching programs with no cases at all (more
precisely, those that have only refutation cases, like `function _ ->
.`), we retain this logic and move it to `toplevel_handler`.

gasche committed:
matching: fix the 'totality' aspect of The Pattern-Matching Bug

gasche committed:
testsuite: an interesting example from Nick Roberts

gasche committed:
matching: track 'global' and 'current' partiality

(no behavior change)

gasche committed:
matching: in mk_failaction_pos, use 'global' partiality to produce final exits

ncik-roberts committed:
new testcases involving polymorphic variants and or-patterns

gasche committed:
matching: no need for final exit on globally-Total matches with empty default

This un-does a small pessimization of #13076, which was analyzed in
details in
https://github.com/ocaml/ocaml/pull/13076#discussion_r1554649410, but
at the time was believed to not affect any program in the wild.

Since then Nick Roberts found instances of this pattern involving
or-patterns and polymorphic variants.

Thanks to the new presentation of partiality information that
distingues the "global" and "current" information, the pessimization
is now easy to undo, as done in this commit.

ncik-roberts reviewed (approved):
I agree with the reasoning behind removing this old heuristic.

The testsuite output needs updating in light of this improvement. Compilation of this function in `testsuite/tests/match-side-effects/check_partial.ml` no longer produces a Match_failure case:

```
let lazy_total : _ * bool t -> int = function
  | ({ contents = _ }, True) -> 0
  | ({ contents = lazy () }, False) -> 12
(* This pattern-matching is in fact total: a Match_failure case is
   not necessary for soundness. *)
```

gasche commented:
I rebased this PR on top of the previous PR's rebase, and fixed the testsuite output, thanks!

(Explanation for other people: the testsuite output recorded a case where no Match_failure case was necessary, but the compiler would conservatively insert one -- which is correct. Removing the check_partial heuristic (replacing it with the finer-grained checks in The Pattern-Matching Bug PRs) makes the compiler more clever about this case, and now there is no Match_failure case anymore. This is a small improvement, and/but both the previous and current test output are correct.)

lpw25 reviewed (approved):
Approving on the basis of @ncik-roberts review

gasche commented:
Thanks!

I clicked "merge" without rebasing the PR (despite the PR being on an older branch than "trunk", it includes some commits that were not in trunk at submission time and have since been merged exactly as-is). Usually I would rebase first, check the CI results, and then merge. It looks like this is working fine -- the CI is being re-run on trunk after the merge, but I think it does this for all merges, even those that do not include commits already in trunk.
